sympy/sets/fancysets.py

```diff
@@ -576,50 +598,62 @@ def reversed(self):
         """
         if not self:
             return self
-        return self.func(
-            self.stop - self.step, self.start - self.step, -self.step)
+        start, stop, step = self.start, self.stop, self.step
+        return self.func(stop - step, start - step, -step)
 
     def _contains(self, other):
+        from sympy.functions.elementary.piecewise import Piecewise
+        from sympy.functions.elementary.integers import ceiling
+        other = _sympify(other)
         if not self:
             return S.false
         if other.is_infinite:
             return S.false
         if not other.is_integer:
             return other.is_integer
         ref = self.start if self.start.is_finite else self.stop
-        if (ref - other) % self.step:  # off sequence
-            return S.false
-        return _sympify(other >= self.inf and other <= self.sup)
+        start, stop, step = self.start, self.stop, self.step
+        inf = Piecewise((start, Ge(step, S.Zero)), (stop - step, True))
+        sup = Piecewise((stop - step, Ge(step, S.Zero)), (start, True))
+        return Piecewise((False, Le(ceiling((stop - start)/step), S.Zero)),
+                         (False, Ne((ref - other) % self.step, S.Zero)),
+                         (True, Ge(other, inf) & Le(other, sup)),
+                         (False, True))
```

Comment:
I don't think Piecewise should be used where the values are non-Expr. If this is a Boolean expression then I think it should probably be a chain of ITE.

Reply:
I will take a look into it. Thanks.